Firestore/core/src/firebase/firestore/local/memory_lru_reference_delegate.cc

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/local/memory_lru_reference_delegate.h"
+
+#include <vector>
+
+#include "Firestore/core/src/firebase/firestore/local/listen_sequence.h"
+#include "Firestore/core/src/firebase/firestore/local/lru_garbage_collector.h"
+#include "Firestore/core/src/firebase/firestore/local/memory_mutation_queue.h"
+#include "Firestore/core/src/firebase/firestore/local/memory_persistence.h"
+#include "Firestore/core/src/firebase/firestore/local/memory_remote_document_cache.h"
+#include "Firestore/core/src/firebase/firestore/local/reference_set.h"
+#include "Firestore/core/src/firebase/firestore/local/remote_document_cache.h"
+#include "Firestore/core/src/firebase/firestore/local/sizer.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+using model::DocumentKey;
+using model::ListenSequenceNumber;
+
+MemoryLruReferenceDelegate::MemoryLruReferenceDelegate(
+    MemoryPersistence* persistence,
+    LruParams lru_params,
+    std::unique_ptr<Sizer> sizer)
+    : persistence_(persistence), sizer_(std::move(sizer)) {
+  gc_ = absl::make_unique<LruGarbageCollector>(this, lru_params);
+
+  // Theoretically this is always 0, since this is all in-memory...
+  ListenSequenceNumber highest_sequence_number =
+      persistence_->query_cache()->highest_listen_sequence_number();
+  listen_sequence_ = absl::make_unique<ListenSequence>(highest_sequence_number);
+}
+
+LruGarbageCollector* MemoryLruReferenceDelegate::garbage_collector() {
+  return gc_.get();
+}
+
+ListenSequenceNumber MemoryLruReferenceDelegate::current_sequence_number() {
+  HARD_ASSERT(current_sequence_number_ != kListenSequenceNumberInvalid,
+              "Asking for a sequence number outside of a transaction");
+  return current_sequence_number_;
+}
+
+void MemoryLruReferenceDelegate::AddInMemoryPins(ReferenceSet* set) {
+  // Technically can't assert this, due to restartWithNoopGarbageCollector (for
+  // now...) FSTAssert(additional_references_ == nil, @"Overwriting additional
```

Comment:
Nit: `FSTAssert` should be on its own line, and probably updated to become `HARD_ASSERT`. `restartWithNoopGarbageCollector` should probably be updated as well.